File: python/cps_utility_console/functions/sensor_stats.py
"""
CPS Utility Console - Sensor Statistics
Author: Dominik Żurek
"""


import math


def compute_stats(samples: list[float]) -> dict[str, float]:
    """
    Compute descriptive statistics for a list of sensor samples.

    Args:
        samples: Non-empty list of float values.

    Returns:
        Dictionary with keys: mean, min, max, stddev.

    Raises:
        ValueError: If samples list is empty.    
    """
    if len(samples) == 0:
        raise ValueError("Samples list cannot be empty.")
    
    mean = sum(samples) / len(samples)
    min_val = min(samples)
    max_val = max(samples)
    variance = sum((x - mean) ** 2 for x in samples) / len(samples)
    stddev = math.sqrt(variance)

    return {
        "mean": mean,
        "min": min_val,
        "max": max_val,
        "stddev": stddev,
    }

File: cpp/src/sensor_stats.cpp
/*
 * CPS Utility Console - Sensor Statistics
 * Author: Dominik Żurek
 */

#include "../include/sensor_stats.hpp"
#include <cmath>
#include <algorithm>
#include <stdexcept>

cps::SensorStats cps::computeStats(const std::vector<float>& samples) {
    if (samples.empty())
        throw std::invalid_argument("Samples list cannot be empty.");

    float mean = 0.0f;
    for (float x : samples) mean += x;
    mean /= static_cast<float>(samples.size());

    float min = *std::min_element(samples.begin(), samples.end());
    float max = *std::max_element(samples.begin(), samples.end());

    float variance = 0.0f;
    for (float x : samples) variance += (x - mean) * (x - mean);
    variance /= static_cast<float>(samples.size());

    return {mean, min, max, std::sqrt(variance)};
}